### ISSUE TITLE
Bug: Data types problem in image.py

### ISSUE THREAD

**Issue by ZheMin-1992** [DeepRegNet/DeepReg#197]
# Issue description
When I do the unit testing for image.py, I met with the following potential issue. 

Inside the dissimilarity_fn function, when calling local_normalized_cross_correlation or ssd, the inputs for those two functions are in tf tensor type.
Otherwise, looking directly into the local_normalized_cross_correlation or ssd  functions, the inputs seem to be numpy array.

The following changes in image.py work for me:
add     "import numpy as np"
add
    "y_true = np.array(y_true)
    y_pred = np.array(y_pred)" after line 15 in the original image.py.

## Type of Issue

- [x] Bug report

**Comment by NMontanaBrown** [comment]
Hi @ZheMin-1992,  can you provide a script that replicates this bug for reproducibility or a set of commands to reproduce this? As well as the environment this has been tested in?

**Comment by ZheMin-1992** [comment]
@NMontanaBrown Thanks. The following script have been ran(seems not work with the current image.py) in the windows subsystem linux(WSL: Ubuntu-18.04).

import numpy as np
import tensorflow as tf
import deepreg.model.loss.image as image
EPS = 1.0e-6
tensor_true = np.array(range(12)).reshape((2, 1, 2, 3))
tensor_pred = np.ones((2, 1, 2, 3))
name_ncc = "lncc"
get_ncc = image.dissimilarity_fn(tensor_true,tensor_pred, name_ncc)
print(get_ncc)

**Comment by NMontanaBrown** [comment]
Hi @ZheMin-1992.The error raised by the following script shows that the two tensors have different types:

`tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-based) was expected to be a int64 tensor but is a double tensor [Op:Mul]`

Doing:
`print(type(tensor_true[0, 0, 0, 0]))
print(type(tensor_pred[0, 0, 0, 0]))`

from your provided code returns:
`<class 'numpy.int64'>
<class 'numpy.float64'>`

What is the behaviour on your local machine if you equalise types? Currently, this script raises no errors for:

`y_true_tensor = tf.ones((2, 4, 4, 4, 4))
y_pred_tensor = tf.zeros((2, 4, 4, 4, 4))

result = image.ssd(y_true_tensor, y_pred_tensor)

result = image.local_normalized_cross_correlation(y_true_tensor, y_pred_tensor)

y_true = np.ones((2, 4, 4, 4, 4))
y_pred = np.zeros((2, 4, 4, 4, 4))

result = image.ssd(y_true, y_pred)
result = image.local_normalized_cross_correlation(y_true, y_pred)`

But if I try to convert to double via `tensor_true = np.array(range(12)).reshape((2, 1, 2, 3)).asarray(np.float64)` it throws similar errors.

In any case, the inputs to all these functions are expected to be tensors. But let's try to figure out this behaviour anyways!

**Comment by ZheMin-1992** [comment]
@NMontanaBrown Thank you so much for your help! Inspired by your comment, I have now first changed the numpy to tf tensor type, and then enforce them all to be tf.float32. In this way, it would work.  The following is the script for your reference. 

```
    import numpy as np
    import tensorflow as tf
    import deepreg.model.loss.image as image

    tensor_true = np.array(range(12)).reshape((2, 1, 2, 3))
    tensor_pred = 0.6 * np.ones((2, 1, 2, 3))
    tensor_true = tf.convert_to_tensor(tensor_true)
    tensor_pred = tf.convert_to_tensor(tensor_pred)
    tensor_true = tf.cast(tensor_true, dtype=tf.float32)
    tensor_pred = tf.cast(tensor_pred, dtype=tf.float32)
    name_ncc = "lncc"
    get_ncc = image.dissimilarity_fn(tensor_true, tensor_pred, name_ncc)
    print(get_ncc）
```

**Comment by NMontanaBrown** [comment]
Hi @ZheMin-1992 sounds, good. In which case, close this issue, and adapt your tests as above. Also would be worth adding docstrings to reflect inputs must be tensor types.

**Comment by mathpluscode** [comment]
@ZheMin-1992 @NMontanaBrown Some small suggestions here to make the code compact

```
    import numpy as np
    import tensorflow as tf
    import deepreg.model.loss.image as image

    tensor_true = np.array(range(12)).reshape((2, 1, 2, 3))
    tensor_pred = 0.6 * np.ones((2, 1, 2, 3))
    tensor_true = tf.convert_to_tensor(tensor_true, dtype=tf.float32)
    tensor_pred = tf.convert_to_tensor(tensor_pred, dtype=tf.float32)
    got = image.dissimilarity_fn(y_true=tensor_true, y_pred=tensor_pred, name="lncc")
    print(got）
```

If you want you can even do

```
    import numpy as np
    import tensorflow as tf
    import deepreg.model.loss.image as image

    tensor_true = tf.convert_to_tensor(np.array(range(12)).reshape((2, 1, 2, 3)), 
                                       dtype=tf.float32)
    tensor_pred = tf.convert_to_tensor(0.6 * np.ones((2, 1, 2, 3)),
                                       dtype=tf.float32)
    got = image.dissimilarity_fn(y_true=tensor_true, y_pred=tensor_pred, name="lncc")
    print(got）
```

I also added some other changes, like specifying the argument name, which makes the code a bit longer but (personally) it's easier to understand. Bonus: your code is more colorful normally in IDE ;)

**Comment by NMontanaBrown** [comment]
Closing as these have been addressed in #198
